{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A Baseline for Models That Leverage The Schwartz-Hearst Algorithm\n",
    "\n",
    "In the Kaggle competition, the Schwartz Hearst Algorithm [(Schwartz & Hearst, 2003)](https://psb.stanford.edu/psb-online/proceedings/psb03/schwartz.pdf) for detecting abbreviations was used by Kaggle Models 1 and 2.\n",
    "\n",
    "Model 2 uses this approach to extract all possible dataset candidates. These are\n",
    "then classified by a binary classifier and the algorithm's other heuristics. So,\n",
    "for at least Kaggle Model 2 this can be used as a baseline for Precision/Recall.\n",
    "This notebook uses specifically model 2's implementation of the algorithm, which\n",
    "is based on the [AllenAI scispacy\n",
    "implementation](https://github.com/allenai/scispacy/blob/cc1a71701f1b83bf81d36c47c271ad5ee1c03d45/scispacy/abbreviation.py),\n",
    "but could have introduced some bugs. \n",
    "\n",
    "For Kaggle Model 1 this is less directly informative because they used the\n",
    "actual AllenAI scispacy code (which may or may not produce different results)\n",
    "and they additionally used a custom algorithm to extract more candidate\n",
    "datasets. Further they used the extracted entities to build up a repository of\n",
    "text snippets to be used for training their model. Their method should be\n",
    "evaluated in another notebook to establish better baselines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from itertools import chain\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from thefuzz import fuzz, process\n",
    "\n",
    "from democratizing_data_ml_algorithms.models.schwartz_hearst_model import SchwartzHearstModel\n",
    "from democratizing_data_ml_algorithms.evaluate.model import evaluate_kaggle_private"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the `scorer` and `processor` arguments are explained in the notebook\n",
    "# `defining_a_match_1.ipynb`\n",
    "\n",
    "evaluation = evaluate_kaggle_private(\n",
    "    SchwartzHearstModel(),\n",
    "    dict(),  # this model doesn't have any configuration params\n",
    "    -1,  # no batching the data\n",
    "    scorer=fuzz.partial_ratio,  # use fuzzy string matching\n",
    "    processor=lambda s: s.lower(),  # convert to lowercase\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`evaluation` is an instance of the `ModelEvaluation` class which has will\n",
    "aggregate some statistics about the model output for us."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "        Model Evaluation:\n",
       "\n",
       "        - Run time: 304.2231593132019 seconds, avg: 0.037730765138683106 seconds per sample\n",
       "        - True Postive Count: 6217, avg: 0.7710529579560957 per sample\n",
       "        - Precision: 0.09721201507356965\n",
       "        - Recall: 0.17481652279054072\n",
       "        "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evaluation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that that the recall is actually much lower than we would hope\n",
    "(~0.17). Let's take a look at what datasets we are most commonly missing. To do\n",
    "so, we look at the `output_statistics` dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>labels</th>\n",
       "      <th>statistics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7bfd8bb51</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'copy-number variants', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7d3e31302</td>\n",
       "      <td>sass|sass data|school and staffing sass data|s...</td>\n",
       "      <td>{'labels': ['sass', 'sass data', 'school and s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3644f959a</td>\n",
       "      <td>foodaps|national household food acquisition an...</td>\n",
       "      <td>{'labels': ['foodaps', 'national household foo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ed3527cf3</td>\n",
       "      <td>database of genotypes and phenotypes|database ...</td>\n",
       "      <td>{'labels': ['database of genotypes and phenoty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>236061129</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8058</th>\n",
       "      <td>3cb8b0e09</td>\n",
       "      <td>database of genotypes and phenotypes|database ...</td>\n",
       "      <td>{'labels': ['database of genotypes and phenoty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8059</th>\n",
       "      <td>324fe1310</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8060</th>\n",
       "      <td>39e66a274</td>\n",
       "      <td>1000 genomes project|1000 genomes project 1000...</td>\n",
       "      <td>{'labels': ['1000 genomes project', '1000 geno...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8061</th>\n",
       "      <td>c411b1b6c</td>\n",
       "      <td>dbgap|american cancer society cancer preventio...</td>\n",
       "      <td>{'labels': ['american cancer society cancer pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8062</th>\n",
       "      <td>990d0ae2d</td>\n",
       "      <td>epa storage and retrieval storet data warehous...</td>\n",
       "      <td>{'labels': ['epa storage and retrieval storet ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8063 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ids                                             labels  \\\n",
       "0     7bfd8bb51                                              dbgap   \n",
       "1     7d3e31302  sass|sass data|school and staffing sass data|s...   \n",
       "2     3644f959a  foodaps|national household food acquisition an...   \n",
       "3     ed3527cf3  database of genotypes and phenotypes|database ...   \n",
       "4     236061129                                              dbgap   \n",
       "...         ...                                                ...   \n",
       "8058  3cb8b0e09  database of genotypes and phenotypes|database ...   \n",
       "8059  324fe1310                                              dbgap   \n",
       "8060  39e66a274  1000 genomes project|1000 genomes project 1000...   \n",
       "8061  c411b1b6c  dbgap|american cancer society cancer preventio...   \n",
       "8062  990d0ae2d  epa storage and retrieval storet data warehous...   \n",
       "\n",
       "                                             statistics  \n",
       "0     {'labels': ['dbgap', 'copy-number variants', '...  \n",
       "1     {'labels': ['sass', 'sass data', 'school and s...  \n",
       "2     {'labels': ['foodaps', 'national household foo...  \n",
       "3     {'labels': ['database of genotypes and phenoty...  \n",
       "4      {'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}  \n",
       "...                                                 ...  \n",
       "8058  {'labels': ['database of genotypes and phenoty...  \n",
       "8059   {'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}  \n",
       "8060  {'labels': ['1000 genomes project', '1000 geno...  \n",
       "8061  {'labels': ['american cancer society cancer pr...  \n",
       "8062  {'labels': ['epa storage and retrieval storet ...  \n",
       "\n",
       "[8063 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# output_statistics is a dataframe with the following columns:\n",
    "# ids: list of ids for the sample\n",
    "# labels: list of labels given by the dataset\n",
    "# statistics: list of dictionaries with the following keys:\n",
    "#   labels: list of set union of input labels and model predictions\n",
    "#   stats: list of with TP, FP, FN fo each of the entries in `labels`\n",
    "evaluation.output_statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats = evaluation.output_statistics\n",
    "all_labels = list(chain(*list(map(lambda x: x[\"labels\"], stats[\"statistics\"].values))))\n",
    "global_stats = list(chain(*list(map(lambda x: x[\"stats\"], stats[\"statistics\"].values))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stats</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>labels</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dbgap</th>\n",
       "      <td>3310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database of genotypes and phenotypes</th>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>program for international student assessment</th>\n",
       "      <td>799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database of genotypes and phenotypes dbgap</th>\n",
       "      <td>519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>schools and staffing survey</th>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and phenotype data and analyses</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and phenotype correlation studies</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and phenotype</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and clinical data</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zulu gwas</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9240 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              stats\n",
       "labels                                             \n",
       "dbgap                                          3310\n",
       "database of genotypes and phenotypes            928\n",
       "program for international student assessment    799\n",
       "database of genotypes and phenotypes dbgap      519\n",
       "schools and staffing survey                     359\n",
       "...                                             ...\n",
       "genotype and phenotype data and analyses          1\n",
       "genotype and phenotype correlation studies        1\n",
       "genotype and phenotype                            1\n",
       "genotype and clinical data                        1\n",
       "zulu gwas                                         1\n",
       "\n",
       "[9240 rows x 1 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_df = pd.DataFrame({\"labels\": all_labels, \"stats\": global_stats})\n",
    "stats_df.loc[stats_df[\"stats\"] == \"FN\", :].groupby(\"labels\").count().sort_values(\n",
    "    \"stats\", ascending=False\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see the most common false negative in this case is `dbgap`, which is commonly\n",
    "written as `dbGaP` and stands for the \"**d**ata**b**ase of **G**enotypes **a**nd\n",
    "**P**henotypes\". Let's look at some documents where Schwartz-Heartz misses the dataset\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>labels</th>\n",
       "      <th>statistics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ed3527cf3</td>\n",
       "      <td>database of genotypes and phenotypes|database ...</td>\n",
       "      <td>{'labels': ['database of genotypes and phenoty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>236061129</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bf7f868aa</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7d9c76c33</td>\n",
       "      <td>1 000 genomes|1 000 genomes reference data|dat...</td>\n",
       "      <td>{'labels': ['1 000 genomes', '1 000 genomes re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>bca78640a</td>\n",
       "      <td>database of genotypes and phenotypes|database ...</td>\n",
       "      <td>{'labels': ['database of genotypes and phenoty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8055</th>\n",
       "      <td>44c7743d6</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'variance stabilizing tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8056</th>\n",
       "      <td>5f8929ee0</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'untranslated regions', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8059</th>\n",
       "      <td>324fe1310</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8060</th>\n",
       "      <td>39e66a274</td>\n",
       "      <td>1000 genomes project|1000 genomes project 1000...</td>\n",
       "      <td>{'labels': ['1000 genomes project', '1000 geno...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8061</th>\n",
       "      <td>c411b1b6c</td>\n",
       "      <td>dbgap|american cancer society cancer preventio...</td>\n",
       "      <td>{'labels': ['american cancer society cancer pr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3141 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ids                                             labels  \\\n",
       "3     ed3527cf3  database of genotypes and phenotypes|database ...   \n",
       "4     236061129                                              dbgap   \n",
       "9     bf7f868aa                                              dbgap   \n",
       "11    7d9c76c33  1 000 genomes|1 000 genomes reference data|dat...   \n",
       "13    bca78640a  database of genotypes and phenotypes|database ...   \n",
       "...         ...                                                ...   \n",
       "8055  44c7743d6                                              dbgap   \n",
       "8056  5f8929ee0                                              dbgap   \n",
       "8059  324fe1310                                              dbgap   \n",
       "8060  39e66a274  1000 genomes project|1000 genomes project 1000...   \n",
       "8061  c411b1b6c  dbgap|american cancer society cancer preventio...   \n",
       "\n",
       "                                             statistics  \n",
       "3     {'labels': ['database of genotypes and phenoty...  \n",
       "4      {'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}  \n",
       "9      {'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}  \n",
       "11    {'labels': ['1 000 genomes', '1 000 genomes re...  \n",
       "13    {'labels': ['database of genotypes and phenoty...  \n",
       "...                                                 ...  \n",
       "8055  {'labels': ['dbgap', 'variance stabilizing tra...  \n",
       "8056  {'labels': ['dbgap', 'untranslated regions', '...  \n",
       "8059   {'labels': ['dbgap', ''], 'stats': ['FN', 'FP']}  \n",
       "8060  {'labels': ['1000 genomes project', '1000 geno...  \n",
       "8061  {'labels': ['american cancer society cancer pr...  \n",
       "\n",
       "[3141 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def model_missed_label(label, row):\n",
    "    if label in row[\"labels\"]:\n",
    "        lbl_idx = row[\"labels\"].index(label)\n",
    "        return row[\"stats\"][lbl_idx] == \"FN\"\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "\n",
    "missed_mask = stats[\"statistics\"].apply(lambda row: model_missed_label(\"dbgap\", row))\n",
    "missed = stats.loc[missed_mask, :]\n",
    "missed"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, the relevant snippet from `ed3527cf3.json`:\n",
    "\n",
    "*The National Institutes of Health (NIH) has established NIH-designated data repositories (e.g., **database of Genotypes and Phenotypes (dbGaP)**, Sequence Read Archive (SRA), NIH Established Trusted Partnerships) for securely storing and sharing controlled-access human data submitted to NIH under the NIH Genomic Data Sharing (GDS) Policy.*\n",
    "\n",
    "This entry is peculiar because the mention is within a set of parenthesis, so it's clear how that affects the algorithm. It is also probably an uncommon occurrence.\n",
    "\n",
    "Next, the relevant snippet from `236061129.json`:\n",
    "\n",
    "*The accession number for data generated from small RNA-sequencing of the maternal and non-maternal plasma samples reported in this paper is **dbGaP**: phs001892.v1.p1 (https://www.ncbi.nlm.nih.gov/gap/).*\n",
    "\n",
    "This doesn't follow the expected pattern and so is an expected false negative.\n",
    "\n",
    "Next, the relevant snippet from `bf7f868aa.json`:\n",
    "\n",
    "*35 All primary sequencing data will be made publicly available in **dbGAP**, accession number phs000759.v1.p1.*\n",
    "\n",
    "This also doesn't follow the expected pattern, so it could be the case that dbgap is so common that referring to it by acronym alone is fine within texts referring to it. \n",
    "\n",
    "Finally here is the snippet from `7d9c76c33.json`:\n",
    "\n",
    "*We evaluate the performance of the proxy algorithm with data sets that were simulated using realistic linkage disequilibrium patterns obtained from the 1,000 Genomes project. Moreover, we successfully applied our approach to Type II diabetes (T2D) GWAS data derived from the **database of Genotypes and Phenotypes (dbGaP)***\n",
    "\n",
    "This seems to be correctly formatted so there is likely an issue with the Schwartz-Hearst Algorithm implementation from Kaggle Model 2.\n",
    "\n",
    "Below we will rerun the analysis, but this time using the Schwartz-Hearst algorithm as implemented by scispacy and AllenAI. Note, that this will take longer because of the time that spacy takes to parse a document, but it may better parse out sentences than the simple implementation from Kaggle Model 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1a1e560a4fd94e9d893d36d9e65d0bcc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=2016), Label(value='0 / 2016'))), …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ryanhausen/repos/democratizingdata-ml-algorithms/src/models/schwartz_hearst_allenai.py:235: UserWarning: [W036] The component 'matcher' does not have any patterns defined.\n",
      "  global_matches = self.global_matcher(doc)\n",
      "/home/ryanhausen/repos/democratizingdata-ml-algorithms/src/models/schwartz_hearst_allenai.py:235: UserWarning: [W036] The component 'matcher' does not have any patterns defined.\n",
      "  global_matches = self.global_matcher(doc)\n",
      "/home/ryanhausen/repos/democratizingdata-ml-algorithms/src/models/schwartz_hearst_allenai.py:235: UserWarning: [W036] The component 'matcher' does not have any patterns defined.\n",
      "  global_matches = self.global_matcher(doc)\n",
      "/home/ryanhausen/repos/democratizingdata-ml-algorithms/src/models/schwartz_hearst_allenai.py:235: UserWarning: [W036] The component 'matcher' does not have any patterns defined.\n",
      "  global_matches = self.global_matcher(doc)\n"
     ]
    }
   ],
   "source": [
    "import src.models.schwartz_hearst_allenai_model as allenai\n",
    "\n",
    "reload(allenai)\n",
    "\n",
    "evaluation_allenai = evaluate_kaggle_private(\n",
    "    allenai.SchwartzHearstModel_AllenAI(model=\"en_core_sci_sm\"),\n",
    "    dict(\n",
    "        char_limit=50000\n",
    "    ),  # limit the amount of the doc that gets processed for memory issues\n",
    "    -1,  # no batching the data\n",
    "    scorer=fuzz.partial_ratio,  # use fuzzy string matching\n",
    "    processor=lambda s: s.lower(),  # convert to lowercase\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "        Model Evaluation:\n",
       "\n",
       "        - Run time: 2594.315016269684 seconds, avg: 0.3217555520612283 seconds per sample\n",
       "        - True Postive Count: 21559, avg: 2.6738186779114472 per sample\n",
       "        - Precision: 0.09690527025508484\n",
       "        - Recall: 0.6106500495680498\n",
       "        "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evaluation_allenai"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the AllenAI implementation improved recall significantly from 0.17 to 0.61. However the computational overhead was significantly higher, from 300 seconds total on one core to 2661 second total over 4 cores. The per document average increased from 0.03 seconds per sample to 0.33 seconds per sample.\n",
    "\n",
    "Let's look at the same attributes of the output to see how things have changed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_allen = evaluation_allenai.output_statistics\n",
    "all_labels_allen = list(\n",
    "    chain(*list(map(lambda x: x[\"labels\"], stats_allen[\"statistics\"].values)))\n",
    ")\n",
    "global_stats_allen = list(\n",
    "    chain(*list(map(lambda x: x[\"stats\"], stats_allen[\"statistics\"].values)))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stats</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>labels</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dbgap</th>\n",
       "      <td>2742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database of genotypes and phenotypes</th>\n",
       "      <td>397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>program for international student assessment</th>\n",
       "      <td>208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>national postsecondary student aid study</th>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database of genotypes and phenotypes dbgap</th>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and clinical data</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and phenotype correlation studies</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and phenotype data and analyses</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype and phenotype database</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zillow transaction and assessment dataset</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4962 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              stats\n",
       "labels                                             \n",
       "dbgap                                          2742\n",
       "database of genotypes and phenotypes            397\n",
       "program for international student assessment    208\n",
       "national postsecondary student aid study        159\n",
       "database of genotypes and phenotypes dbgap      146\n",
       "...                                             ...\n",
       "genotype and clinical data                        1\n",
       "genotype and phenotype correlation studies        1\n",
       "genotype and phenotype data and analyses          1\n",
       "genotype and phenotype database                   1\n",
       "zillow transaction and assessment dataset         1\n",
       "\n",
       "[4962 rows x 1 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_df_allen = pd.DataFrame({\"labels\": all_labels_allen, \"stats\": global_stats_allen})\n",
    "stats_df_allen.loc[stats_df_allen[\"stats\"] == \"FN\", :].groupby(\n",
    "    \"labels\"\n",
    ").count().sort_values(\"stats\", ascending=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's compare the top 5 between the two. An empty space indicates its not\n",
    "in the top 5.\n",
    "\n",
    "|dataset| Model 2 | AllenAI |\n",
    "|-------|---------|---------|\n",
    "|dbgap                                      |3,310|2,742|\n",
    "|database of genotypes and phenotypes       |928  |397  |\n",
    "|program for interational student assessment|799  |208  |\n",
    "|database of genotypes and phenotypes dbgap |519  |146  | \n",
    "|schools and staffing survey                |359  |     |\n",
    "|national postsecondary student aid study   |     |159  |\n",
    "\n",
    "Let's look at some of the missed examples to see if this approach also seems to\n",
    "be missing things it should catch. First let's check `dbgap`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>labels</th>\n",
       "      <th>statistics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7bfd8bb51</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'fl', 'Whole-exome sequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>236061129</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'representative of relati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bf7f868aa</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'hprt', 'MEFs', 'N-methyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>de28ee4d4</td>\n",
       "      <td>1000 genomes project|clinvar|database of genot...</td>\n",
       "      <td>{'labels': ['database of genotypes and phenoty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>d58f1872e</td>\n",
       "      <td>gtex|genotype tissue expression gtex consortiu...</td>\n",
       "      <td>{'labels': ['gtex', 'genotype tissue expressio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8056</th>\n",
       "      <td>5f8929ee0</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'GWAS', 'gene B', 'genome...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8058</th>\n",
       "      <td>3cb8b0e09</td>\n",
       "      <td>database of genotypes and phenotypes|database ...</td>\n",
       "      <td>{'labels': ['database of genotypes and phenoty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8059</th>\n",
       "      <td>324fe1310</td>\n",
       "      <td>dbgap</td>\n",
       "      <td>{'labels': ['dbgap', 'loss of heterozygosity',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8060</th>\n",
       "      <td>39e66a274</td>\n",
       "      <td>1000 genomes project|1000 genomes project 1000...</td>\n",
       "      <td>{'labels': ['lhs', 'lhs study', 'lung health s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8061</th>\n",
       "      <td>c411b1b6c</td>\n",
       "      <td>dbgap|american cancer society cancer preventio...</td>\n",
       "      <td>{'labels': ['american cancer society cancer pr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2607 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ids                                             labels  \\\n",
       "0     7bfd8bb51                                              dbgap   \n",
       "4     236061129                                              dbgap   \n",
       "9     bf7f868aa                                              dbgap   \n",
       "18    de28ee4d4  1000 genomes project|clinvar|database of genot...   \n",
       "21    d58f1872e  gtex|genotype tissue expression gtex consortiu...   \n",
       "...         ...                                                ...   \n",
       "8056  5f8929ee0                                              dbgap   \n",
       "8058  3cb8b0e09  database of genotypes and phenotypes|database ...   \n",
       "8059  324fe1310                                              dbgap   \n",
       "8060  39e66a274  1000 genomes project|1000 genomes project 1000...   \n",
       "8061  c411b1b6c  dbgap|american cancer society cancer preventio...   \n",
       "\n",
       "                                             statistics  \n",
       "0     {'labels': ['dbgap', 'fl', 'Whole-exome sequen...  \n",
       "4     {'labels': ['dbgap', 'representative of relati...  \n",
       "9     {'labels': ['dbgap', 'hprt', 'MEFs', 'N-methyl...  \n",
       "18    {'labels': ['database of genotypes and phenoty...  \n",
       "21    {'labels': ['gtex', 'genotype tissue expressio...  \n",
       "...                                                 ...  \n",
       "8056  {'labels': ['dbgap', 'GWAS', 'gene B', 'genome...  \n",
       "8058  {'labels': ['database of genotypes and phenoty...  \n",
       "8059  {'labels': ['dbgap', 'loss of heterozygosity',...  \n",
       "8060  {'labels': ['lhs', 'lhs study', 'lung health s...  \n",
       "8061  {'labels': ['american cancer society cancer pr...  \n",
       "\n",
       "[2607 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def model_missed_label(label, row):\n",
    "    if label in row[\"labels\"]:\n",
    "        lbl_idx = row[\"labels\"].index(label)\n",
    "        return row[\"stats\"][lbl_idx] == \"FN\"\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "\n",
    "missed_mask = stats_allen[\"statistics\"].apply(\n",
    "    lambda row: model_missed_label(\"dbgap\", row)\n",
    ")\n",
    "missed = stats_allen.loc[missed_mask, :]\n",
    "missed"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The mention candidate in `7bfd8bb51.json`:\n",
    "\n",
    "*Whole-exome sequencing data have been deposited in the **dbGaP** database (http://www. ncbi.nlm.nih.gov/gap) under the accession number phs000474.v3.p2.*\n",
    "\n",
    "This doesn't follow the expected pattern, so it is an expected false negative. The mention in document `236061129.json`:\n",
    "\n",
    "*The accession number for data generated from small RNA-sequencing of the maternal and non-maternal plasma samples reported in this paper is **dbGaP**: phs001892.v1.p1 (https://www.ncbi.nlm.nih.gov/gap/).*\n",
    "\n",
    "This also an expected false negative. The mention in document `bf7f868aa.json` is the following:\n",
    "\n",
    "*35 All primary sequencing data will be made publicly available in **dbGAP**, accession number phs000759.v1.p1.*\n",
    "\n",
    "This is also an expected false negative. Then mention in document `de28ee4d4.json` looks like:\n",
    "\n",
    "*Whole-exome sequencing data have been submitted to the **Database of Genotypes and Phenotypes (dbGaP; https://www.ncbi.nlm.nih.gov/gap)** per NIH study protocol and patient consent (accession number phs001232.v1.p1).*\n",
    "\n",
    "This presents an interesting example of a false negative. It's nearly the correct format, but has what is not a completely uncommon format for multiple citations for a single reference **LONG FORM (ACRONYM; CITE)**. I wonder if this should be added in how the parser looks for candidates?\n",
    "\n",
    "How do the precision and recall change if we exlude `dbgap` from the calculation?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "precision 0.09316747759051806\n",
      "recall 0.6522783290147254\n"
     ]
    }
   ],
   "source": [
    "non_dbgap = stats_df_allen.loc[stats_df_allen[\"labels\"] != \"dbgap\", :]\n",
    "\n",
    "tp = (non_dbgap[\"stats\"] == \"TP\").sum()\n",
    "fp = (non_dbgap[\"stats\"] == \"FP\").sum()\n",
    "fn = (non_dbgap[\"stats\"] == \"FN\").sum()\n",
    "print(\"precision\", tp / (tp + fp))\n",
    "print(\"recall\", tp / (tp + fn))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "### Summary\n",
    "\n",
    "This notebook explores setting a baseline for all models that use\n",
    "Schwartz-Hearst algorithm for extracting datasets from context in the private\n",
    "kaggle. From the experiments we can see that although slower, the AllenAI \n",
    "implementation of the Schwartz-Hearst algorithm gives a higher recall than the\n",
    "faster implementation from model 2. \n",
    "\n",
    "The best recall (at least in using the private kaggle dataset) is $\\approx$ 0.61.\n",
    "\n",
    "If we exclude `dbgap` from the calculation the we best recall we can have is \n",
    "$\\approx$ 0.65.\n",
    "\n",
    "So if the distribution of the dataset references in the private kaggle vaidation\n",
    "set is similar to the distribution of dataset references in general then I other\n",
    "methods like Kaggle Model 1's approach will likely be needed to improve the\n",
    "recall."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deomcratizing-data",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "86de337ddf643eb96c825eddf188b9d07b329ad3e6f022c2cbb07a9b50fdd5ab"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
